src/sundials/sundials_adaptcontroller.c

```diff
@@ -0,0 +1,173 @@
+  if (ops == NULL) { free(C); return(NULL); }
+
+  /* initialize operations to NULL */
+  ops->gettype              = NULL;
+  ops->destroy              = NULL;
+  ops->reset                = NULL;
+  ops->estimatestep         = NULL;
+  ops->setdefaults          = NULL;
+  ops->write                = NULL;
+  ops->seterrorbias         = NULL;
+  ops->updateh              = NULL;
+  ops->space                = NULL;
+
+  /* attach ops and initialize content to NULL */
+  C->ops     = ops;
+  C->content = NULL;
+  C->sunctx  = sunctx;
+
+  return(C);
+}
+
+
+/* -----------------------------------------------------------------
+ * Free a generic SUNAdaptController (assumes content is already empty)
+ * ----------------------------------------------------------------- */
+
+void SUNAdaptController_FreeEmpty(SUNAdaptController C)
+{
+  if (C == NULL)  return;
+
+  /* free non-NULL ops structure */
+  if (C->ops)  free(C->ops);
+  C->ops = NULL;
+
+  /* free overall SUNAdaptController object and return */
+  free(C);
+  return;
+}
+
+
+/* -----------------------------------------------------------------
+ * Required functions in the 'ops' structure for non-NULL controller
+ * ----------------------------------------------------------------- */
+
+SUNAdaptController_Type SUNAdaptController_GetType(SUNAdaptController C)
+{
+  if (C == NULL) { return SUN_ADAPTCONTROLLER_NONE; }
+  return(C->ops->gettype(C));
+}
+
+/* -----------------------------------------------------------------
+ * Optional functions in the 'ops' structure
+ * ----------------------------------------------------------------- */
+
+int SUNAdaptController_Destroy(SUNAdaptController C)
+{
+  if (C == NULL) return(SUNADAPTCONTROLLER_SUCCESS);
+
+  /* if the destroy operation exists use it */
+  if (C->ops)
+    if (C->ops->destroy) { return(C->ops->destroy(C)); }
+
+  /* if we reach this point, either ops == NULL or destroy == NULL,
+     try to cleanup by freeing the content, ops, and matrix */
+  if (C->content) { free(C->content); C->content = NULL; }
+  if (C->ops) { free(C->ops); C->ops = NULL; }
+  free(C); C = NULL;
+
+  return(SUNADAPTCONTROLLER_SUCCESS);
+}
+
+int SUNAdaptController_EstimateStep(SUNAdaptController C, sunrealtype h, int p,
+                                    sunrealtype dsm, sunrealtype* hnew)
+{
+  int ier = 0;
+  *hnew = h;   /* initialize output with identity */
+  if (C == NULL) { return ier; }
+  if (C->ops->estimatestep)
+  {
+    ier = C->ops->estimatestep(C, h, p, dsm, hnew);
+  }
+  return(ier);
+}
+
+
+int SUNAdaptController_Reset(SUNAdaptController C)
+{
+  int ier = 0;
+  if (C == NULL) { return ier; }
+  if (C->ops->reset) { ier = C->ops->reset(C); }
+  return(ier);
+}
+
+int SUNAdaptController_SetDefaults(SUNAdaptController C)
+{
+  int ier = 0;
+  if (C->ops->setdefaults) { ier = C->ops->setdefaults(C); }
+  return(ier);
+}
+
+int SUNAdaptController_Write(SUNAdaptController C, FILE* fptr)
+{
+  int ier = 0;
+  if (C == NULL) { return ier; }
```

Comment:
```suggestion
  if (C == NULL || fptr == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

Reply:
I don't see why this check on `fptr` needs to be duplicated everywhere. If it is here then there is no reason to do it redundantly in each implementation (as requested).